v5: Move from JS input button group toggling to all CSS

mdo opened the pull request:
Takes #25185 and starts work on the docs and JS to drop code that was previously handling active states.

/cc @ysds 

- [x] make sure the CSS changes are OK
- [x] **make sure the JS changes are OK, tests pass, visual tests are also OK**
- [ ] Check if we need to revisit bundlesize limits
- [x] Finally, **make sure all functionality we had before is still there, especially accessibility-wise** /CC @patrickhlauke

mdo committed:
Cherry-pick from ysds

mdo committed:
Update docs to match new markup, drop JS

mdo commented:
Superseded by #28463.

ysds commented:
The CSS LGTM👍  IMO, it is better to move the documents to the button group page.

XhmikosR committed:
Update button.js

XhmikosR commented:
Do not merge yet, @mdo, until 1) we make sure we don't lose any functionality, and 2) JS and its tests/including visual are fine.

ysds commented on site/docs/4.3/components/buttons.md:
```diff
@@ -125,32 +125,36 @@ Add `data-toggle="button"` to toggle a button's `active` state. If you're pre-to
-  </label>
+<div class="btn-group">
+  <div class="btn-group">
+    <input type="checkbox" class="btn-group-input" name="btnGroupRadio" id="btnGroupRadio1" checked>
```
This and the remain two lines should be `type="radio"`.

ysds commented on scss/_button-group.scss:
```diff
@@ -148,16 +150,8 @@
+.btn-group-input {
+  position: absolute;
+  clip: rect(0, 0, 0, 0);
+  pointer-events: none;
```
Need to remove label's bottom margin.
```suggestion
  pointer-events: none;

  ~ .btn {
    margin-bottom: 0; // Override default `<label>` value
  }
}
```

ysds committed:
Update scss/_button-group.scss

Co-Authored-By: mdo <otto@github.com>

mdo committed:
Fix input types

XhmikosR commented on js/src/button.js:
```diff
@@ -65,56 +62,10 @@ class Button {
-    if (triggerChangeEvent) {
-      this._element.classList.toggle(ClassName.ACTIVE)
-    }
+    this._element.classList.toggle(ClassName.ACTIVE)
```
So, I guess we don't need the active class any longer, right?

I think we'll need to find another way to add the `aria-pressed` attribute.

mdo replied:
Yeah, we might be able to nix all the JS—can anyone confirm that? Only reason I could see us leaving some JS for a non-input implementation is quickly toggling active on/off, but I feel like most folks could do that themselves with very little code and their own syntax.

XhmikosR commented:
And who's going to handle aria pressed?

On Tue, Mar 12, 2019, 22:48 Mark Otto <notifications@github.com> wrote:

> *@mdo* commented on this pull request.
> ------------------------------
>
> In js/src/button.js
> <https://github.com/twbs/bootstrap/pull/28463#discussion_r264876819>:
>
> >
>
> -    if (triggerChangeEvent) {
>
> -      this._element.classList.toggle(ClassName.ACTIVE)
>
> -    }
>
> +    this._element.classList.toggle(ClassName.ACTIVE)
>
>
> Yeah, we might be able to nix all the JS—can anyone confirm that? Only
> reason I could see us leaving some JS for a non-input implementation is
> quickly toggling active on/off, but I feel like most folks could do that
> themselves with very little code and their own syntax.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/twbs/bootstrap/pull/28463#discussion_r264876819>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAVVtVsLpVRuedaqHbVJKgo5IdqJ9DvSks5vWBK3gaJpZM4bp4Wd>
> .
>

ysds commented on scss/_button-group.scss:
```diff
@@ -148,16 +150,13 @@
+    margin-bottom: 0; // Override default `<label>` value
   }
 }
+}
```
Sorry, my previous suggestion added a duplicate '}' 🙏 

```suggestion
```

ysds commented:
It will be difficult to solve the #25122 and #27600 issues as long as we use `<input type="checkbox">` and `<input type="radio">` to handle the `aria-pressed`.

According to @scottaohara's [explanation](https://scottaohara.github.io/a11y_styled_form_controls/src/radio-button--pill/), there is a radio button group implementation pattern without `aria-pressed`.

Also, his comment at the end of the article suggests a grouping of `<button>`s. It is probably something like the [WAI-ARIA practice](https://www.w3.org/TR/wai-aria-practices-1.1/examples/toolbar/toolbar.html).

> Alternatively, a grouping of button elements that are aware of each other's state, and utilize aria-pressed="true/false" to indicate their own state, may be more appropriate and understandable to users.

So in terms of Pure CSS and accessibility, we'll need decide to which one:

- A. Implement by Pure CSS (`<input>`) without `aria-pressed`
- B. Implement with `<button>` using JavaScript to control the aria attributes

scottaohara commented:
> Also, his comment at the end of the article suggests a grouping of `<button>`s. It is probably something like the [WAI-ARIA practice](https://www.w3.org/TR/wai-aria-practices-1.1/examples/toolbar/toolbar.html).

that could be one method but not necessarily the only method, in that a grouping could be just a `group` of buttons, rather than a `toolbar`. More importantly though, each would provide different user expectations to the manner in which they were keyboard accessible, as would checkboxes and radios.

As I'm not deeply familiar with what it is you're trying to solve for here, and honestly i'm a bit confused reading through just this thread with the number of different elements being referenced, I'd definitely defer to @patrickhlauke's guidance.

patrickhlauke commented:
coming in late on this...so: skimming over the changes, it looks like we're going for clean CSS version here? just layering over texisting `<input>` elements? in that case, if those inputs are there and focusable (and not suppressed), there won't be need for any JS or aria or anything, unless I'm missing something.

XhmikosR commented:
Yeah, the idea is to use only CSS for buttons. If we don't need to handle aria, I can update the branch dropping more stuff.

patrickhlauke commented:
ah, hang on...i think the confusion here is because there's two separate things going on...there's the "checkboxes and radio buttons that look like a toolbar/button strip" https://getbootstrap.com/docs/4.3/components/buttons/#checkbox-and-radio-buttons (which is what wouldn't need any JS nor `aria-pressed` or anything, and can be solved purely with CSS), and then there's the "set of buttons that acts as mutually exclusive toggles" https://getbootstrap.com/docs/4.3/components/button-group/ (which DO need `aria-pressed` to denote their current toggle state)

I'll be honest, i find these two things related, but confusing. Most confusing that the checkbox/radio ones are under the "Buttons" documentation - should they perhaps be under their own component topic for checkbox/radio button group?

XhmikosR commented:
Hmm then I'm also lost. Not sure how to proceed. Maybe we should rename `button` to something else and only keep what we need in JS?

patrickhlauke commented:
incidentally, am i missing something, or are the groups in https://getbootstrap.com/docs/4.3/components/button-group/ not actually doing anything fancy JS wise at the moment? are they supposed to have aria-pressed and act like sets of radio buttons or not? the "button plugin" mention on there simply points back to the buttons page...so seems we're making some kind of circular reference here.

worth taking a step back and deciding overall what we want. certainly the easiest part is:

- we can style checkboxes and radio buttons with CSS to appear button-like; behind the scenes, it's still just an `<input>` and a `<label>`, and uses CSS - no need for JS/ARIA. this is probably a special case of form element styling, not necessarily/logically something that should fall under "button" component documentation?
- for `<button>` elements to act as on/off type toggles, they need JS and `aria-pressed` handling

Now, button groups are like an extended version of the single button that acts as a toggle. A group can be mutually exclusive (acting like a set of radio buttons) or not (like a toolbar in a wysiwyg tool, where each on/off toggle is independent - set bold, or italic, or underline, independently. Those, again, need JS. and for the mutually exclusive capability, they need JS beyond just the pure toggle JS functionality, as it needs to not only toggle `aria-pressed` on what the user selected, but also unset it on any others in the same group.

Conceptually then, the single toggle button plugin thing, and button groups, probably should go into their own component documentation page together, as they're thematically related (groups build on the single toggle). Whereas the implementation using radio/checkbox is just a differently styled set of form controls?

ysds committed:
Update scss/_button-group.scss

Co-Authored-By: MartijnCuppens <martijn.cuppens@gmail.com>

mdo commented:
Answering some questions here...

- Buttons in button groups are not intended to include JS functionality. Button groups are a component for controlling layout of buttons. None of these buttons _need_ JS from us—anyone can write their own JS to toggle the `.active` class.
- My proposal earlier is to drop our button JS plugin and use only `<input>`s and `<label class="btn">` combos.
- Given that, and per @ysds's feedback, we can then move the input/label examples to the button group docs page.

If all that's true and agreeable, it sounds like we don't need any JS here or any `aria` attributes. If that doesn't sound right, let's reframe the conversation around "how do we remove all this JS?" and answer any questions that come from that.

patrickhlauke commented:
@mdo thanks for clarifying. so yes, basically removing the "Add on optional JavaScript radio and checkbox style behavior with our buttons plugin." sentence, and just making button groups a visual style? And just having actual radio/checkbox inputs to act as such, without the button plugin? In that case yes, no need anymore for the JS nor ARIA. and it would remove the confusion.

Would be good to provide some clearer examples for button groups, as currently some of them make it still look like they're supposed to be some mutually exclusive radios. But that can be revisitied in a separate PR.

patrickhlauke commented:
related, if we want to at least fix current keyboard issue in v4 https://github.com/twbs/bootstrap/pull/28565

patrickhlauke committed:
Remove unit tests for faked radio button/checkbox buttons

patrickhlauke committed:
Add unit test for focus class handling

patrickhlauke committed:
Initial update of button documentation to match new reality

As we now just use real radio/checkbox, there's no weird "only updated on click" clause. And having a pre-checked button is a case of just checking the input.

patrickhlauke committed:
Add handling of disabled buttons

patrickhlauke committed:
Unit test for disabled buttons and for btn lacking initial aria-pressed

patrickhlauke committed:
Remove obsolete mention of super-powered/JS functionality for button groups

patrickhlauke committed:
Reorganise/reword button documentation

- Move the checkbox/radio button section out of the "Button plugin" part, as it now does not rely on the plugin anymore.
- Reword the intro to the checkbox/radio button section, as it's far more focused and specific now, rather than "just add styles to other elements to make them like a checkbox/radio".
- Reword the intro to the button plugin section, as it's now purely about providing toggles (incidentally...we may want to consider renaming the plugin itself to "toggle plugin?")

patrickhlauke committed:
Remove obsolete unit test

This is not documented behavior. This structure makes no sense if the button plugin now just acts as a way to turn a button into a toggle and nothing more. also, this test was flawed as it toggled the toggle function instead of simulating a click...

patrickhlauke committed:
Simplify button.js selector and add comment for click behavior

- As we're now just supporting `.btn` elements with `[data-toggle="button"]`, and none of those faked button groups etc, the selector can be simplified (no more carrot ?!).
- Added a comment (mainly for myself?) to clarify why we're still trying to find the closest button to the event.target - for cases like `<button ...><i>foo</i></button>` where otherwise the `event.target` is the `<i>` and so wouldn't work/match the class check. This is due to the way the whole EventHandler/delegation work ... if the events were tied directly to the elements, rather than delegated, would just need to handle the `this` and be done with it...

patrickhlauke committed:
Use proper simulated click in one test, move related unit tests closer together

patrickhlauke committed:
Redo button visual test

- remove the now unnecessary test for faked checkbox/radio button groups, aas the plugin now only handles toggles
- expand the text of the single toggle to make it clear what's expected
- add a disabled toggle that shouldn't be pressable
- add a button that initially is missing `aria-pressed`

patrickhlauke committed:
Change unit tests to test actual simulated click

This not only checks that the toggle function itself executes correctly, but that the event listener/handling is working as it should

patrickhlauke committed:
Reintroduce one unit test for bootstrapButton('toggle')

- needed for 100% statement code coverage of button.js

patrickhlauke committed:
Merge branch 'master' into v5-input-btn-group

patrickhlauke committed:
Remove unnecessary autocomplete="off"

`<button>` elements aren't autocompleted by browsers (as they don't have a particular state/value by themselves, we're
faking/adding it ourselves with `aria-pressed`), so these explicit `autocomplete="off"` attributes do nothing

patrickhlauke committed:
Tweak focus/blur test

IE11 doesn't just let you blur something without then focusing something else. In real-world situation, this would be the case - you'd move focus away and onto something different (even just the main document), so testing this way is closer to reality than artificially firing blur()

patrickhlauke committed:
Expand button group description

Also mention vertical column/stack

patrickhlauke committed:
Further tweak to unit test

turns out IE11 doesn't always/properly fire focus/blur event handlers when elements triggered purely programmatically with `focus()` / `blur()`.

patrickhlauke committed:
Reintroduce element checks for focus/blur, tweak unit test again

IE11 still failing inexplicably in unit test but not in real-world usage.

patrickhlauke committed:
Remove bespoke focus styling

Sudden realisation that because we now only really support properly focusable elements for toggles, they would already get the standard `:focus` anyway. No need for custom addition of `.focus` programmatically (which is also causing issues with the unit testing in IE11 as it doesn't fire the focus/blur handlers correctly when triggered programmatically, it seems)

patrickhlauke committed:
Expand automated and manual tests

- for automated tests, make sure we check faked buttons (using `<div>` and `<a href="#">`) as well - there's generally no reason for authors to use these over actual `<button>` elements, but hey...some will, and our plugin should handle this
- added these fake buttons to the visual/manual unit test as well, with custom keyboard handling and a bit more explanation and styling

patrickhlauke commented:
one thing i did notice: when using real `<button>` (for toggles and elsewhere), IE11 does the "content of the button moves down and right a few pixels when pressed" dance. surprised this wasn't normalised away...was it not before?

patrickhlauke committed:
Expand checkbox/radio button explanation

Explicitly mention that the structure needs to be just so, because of the sibling selector used

patrickhlauke commented:
https://deploy-preview-28463--twbs-bootstrap.netlify.com/docs/4.3/components/buttons/#checkbox-and-radio-buttons work and look as they should
https://deploy-preview-28463--twbs-bootstrap.netlify.com/docs/4.3/components/buttons/#button-plugin works great too
and both automated and visual tests are all ok
so, at this stage - barring any fundamental changes or concerns that maybe i've gone too far in slashing things apart - this seems functionally good from my point of view...

patrickhlauke committed:
Add automated and visual test for disabled <a href=#"> faked button

patrickhlauke committed:
Tighten and combine related unit tests

patrickhlauke committed:
Stupid typos/unnecessary ids in visual test

patrickhlauke commented:
I'll see if I can fix up the old-school button in IE11 issue with the text dropping down when pressed/clicked...after that, this may be ready to properly review

patrickhlauke commented:
> I'll see if I can fix up the old-school button in IE11 issue with the text dropping down when pressed/clicked...after that, this may be ready to properly review

doesn't seem to be any clean fix (some stuff i've seen involves wrapping all the content of the button in a `<span>` and applying some CSS translate stuff, or setting `position:relative` ... but in any case, stuff that would need extra markup that I wouldn't want to require from authors just so that the IE11 fix works). as we've not normalized/reset this in the past, i think then that that can stay as is, and this PR is ready for review. /cc @twbs/css-review @twbs/js-review

patrickhlauke commented:
just skimming over the previous early conversation here, there was the idea of moving the checkbox/radio stuff to button group rather than leaving them in the button docs page. i see the point to an extent, but i think more logically the now amended checkbox/radio stuff goes well in button docs page, as it shows how you can make checkboxes/radio buttons that look like buttons. the fact that they're in a button group is incidental

patrickhlauke committed:
Rename btn-group-input to just btn-input

The class isn't necessarily just for when you want to use these in a group. You can have these just as a single checkbox, or a set of button-like radio buttons NOT in a button group. Makes more
sense to generalise the name.

patrickhlauke commented:
actually, few more tiny tweaks that make this even more explicit...

patrickhlauke committed:
Use + next-sibling combinator not ~ subsequent-sibling combinator

with ~ it would apply to one or more subsequent siblings. we only want it to affect the sibling immediately after the input, as there's no need to have any further markup in between/around the
label

patrickhlauke committed:
Update/extend buttons and button-group docs

- for buttons, just show the clean radio/checkbox buttons
- then in button-group, show how they can be styled into a seamless group

patrickhlauke committed:
Move btn-input to more logical place, fix button group styles to allow for radio/checkbox buttons

patrickhlauke commented:
"tiny" tweaks ended up simplifying the required structure for the new button-like radio/checkbox buttons (now just needs the input and label adjacent, no wrapper around it), more clearly showing the difference between having a series of these, and having them in a seamless-looking button group. docs now show this nicely/clearly as well

patrickhlauke committed:
Merge branch 'master' into v5-input-btn-group

XhmikosR committed:
Update button.html

XhmikosR commented on site/content/docs/4.3/components/progress.md:
```diff
@@ -127,7 +127,7 @@ The striped gradient can also be animated. Add `.progress-bar-animated` to `.pro
   <div class="progress">
     <div class="progress-bar progress-bar-striped" role="progressbar" aria-valuenow="75" aria-valuemin="0" aria-valuemax="100" style="width: 75%"></div>
   </div>
-  <button type="button" class="btn btn-secondary mt-3" data-toggle="button" id="btnToggleAnimatedProgress" aria-pressed="false" autocomplete="off">
```
Maybe this change shouldn't be in this PR?

XhmikosR replied:
Or am I wrong... I thought you had another PR for removing autocomplete.

patrickhlauke replied:
think it makes sense to keep it here as it's related to toggle buttons and how they're then presented/used throughout the docs, rather than leave it in and remove it later (the other branch is still in limbo as it's not quite clear if Firefox still has that bug that prompted it in the first place or not)

ysds commented:
> now just needs the input and label adjacent, no wrapper around it

Oh, I intentionally wrapped those in `.btn-group` for margin control. https://codepen.io/anon/pen/QRXYgm Please see the left edge (-1px margin).

And it also available of labeling without `id/for`.

patrickhlauke commented:
> Oh, I intentionally wrapped those in `.btn-group` for margin control. https://codepen.io/anon/pen/QRXYgm Please see the left edge (-1px margin).

Can that be fixed in the current approach?

> And it also available of labeling without `id/for`.

Example?

ysds commented:
e.g.
```html
<label class="btn-group">
  <input type="checkbox" class="btn-group-input">
  <span class="btn btn-primary">Single toggle</span>
</label>
```

> Can that be fixed in the current approach?

The current negative margins and border radius control are very complex:

![image](https://user-images.githubusercontent.com/4065765/59141515-22752a80-89e9-11e9-96d5-2091edb5fbce.png)

Current negative margins and border radius control are very complex.

It is not easy to check that the current approach has no problem in various markup by various users. The `.btn-group` wrappers will be a safe approach for that.

patrickhlauke commented:
then i suggest the classnames get changed. it's not a button group, it's a custom button wrapper. i'll have a think

patrickhlauke commented:
> Current negative margins and border radius control are very complex.

would need some testing, but I'm sure something like

```
.btn-group .btn-input + .btn:first-of-type { margin-left: $btn-border-width }
```

of similar could do it.

Perhaps it would be good offering this way (using just the input adjacent label) as an option, and also providing a new class `.btn-input-wrapper` ? I just feel `.btn-group` is misleading, as you're not wrapping it around a group, but around a single checkbox/radio `<input>` and its related label text element (either a `<label>` or a `<span>` or whatever, like in your example. Thoughts?

patrickhlauke committed:
Add new `button-input-wrapper` wrapper, expand CSS, expand docs

- `button-input-wrapper` Makes this arguably more contained/flexible, less prone to problems with varied markup authors may have
- expanded the CSS to cover all extra styling required, including correct focus/hover/active styles and button group additions/tweaks
- expanded docs considerably to give more examples or these checkbox/radio button buttons in action - would love to use outline variants, but
it appears our outline buttons lack distinction between hovered and active/checked, making them useless for now. to be addressed in a separate
PR, and then revisit this with nicer outline buttons that really show off things like button groups with checkboxes/radios nicely

patrickhlauke commented:
@ysds @mdo see what you think of this approach...a bit more involved than i originally anticipated, but essentially adds a `.btn-input-wrapper` (instead of piggy-backing inappropriately on `.btn-group`). expanded the docs to show various examples, both with explicit (`for`/`id`) and implicit `<label>` elements. also tweaks the toggle example slightly to show both pre-toggled and regular toggle.

see https://deploy-preview-28463--twbs-bootstrap.netlify.com/docs/4.3/components/buttons/#checkboxes-and-radio-buttons / https://deploy-preview-28463--twbs-bootstrap.netlify.com/docs/4.3/components/buttons/#toggle-states / https://deploy-preview-28463--twbs-bootstrap.netlify.com/docs/4.3/components/button-group/#checkbox-and-radio-button-groups and the added radio button example for vertical variant https://deploy-preview-28463--twbs-bootstrap.netlify.com/docs/4.3/components/button-group/#vertical-variation

patrickhlauke commented:
as mentioned, would have loved to use the outline variants for these examples (particularly for the button group examples), but it appears currently we lack visual differentiation between hover and active on those (by design it seems), which makes them horribly confusing. will look at addressing that in a separate PR, after which we can revisit the examples here, but that shouldn't hold up review of this one.

patrickhlauke commented:
late night realisation ... these radio/checkbox style buttons lack styles for disabled inputs. putting this back in the oven for a bit longer, should find some time tomorrow to add those as well.

patrickhlauke committed:
Add disabled variants to the checkbox, radio button and toggle button examples

though styles aren't done for these yet, coming next...

patrickhlauke committed:
Add missing :disabled styles for checkbox/radio button buttons

patrickhlauke committed:
Expand examples of buttons further

Just to prove the point that color, outline, size can all be changed here as well

patrickhlauke committed:
Remove generic focus for checkbox/radio .btn

This ended up too strong and overrode the box shadow from the color/outline variations

patrickhlauke commented:
ok...the checkbox/radio button buttons now have proper handling for disabled input inside, and i fixed up a few stragglers like the box shadow not being correct. lastly, added extensive examples to both the buttons and button group docs page to show variations/uses. @mdo @ysds thoughts?

patrickhlauke committed:
Move callout to more logical place in button-group doc

patrickhlauke committed:
Merge branch 'master' into v5-input-btn-group

mdo commented:
Trying to figure out the HTML and CSS here... I'd like to avoid adding tons of new selectors to account for this. Might be impossible, but I'm not giving up just yet. 😬 Starting a Codepen to capture some things with a modified version of our compiled CSS: https://codepen.io/emdeoh/pen/agZGqa?editors=1100.

patrickhlauke commented:
> Trying to figure out the HTML and CSS here... I'd like to avoid adding tons of new selectors to account for this. Might be impossible, but I'm not giving up just yet. 😬 Starting a Codepen to capture some things with a modified version of our compiled CSS: https://codepen.io/emdeoh/pen/agZGqa?editors=1100.

fundamentally, it's confusing to just piggy-back onto `btn-group` for what is not an actual button group, but rather a button-like input. i'd take a few extra selectors for clarity over a bastardisation of btn-group, particularly since these constructs aren't just button groups, but live conceptually under the button component itself (but then CAN be turned into an actual group).

patrickhlauke commented:
i.e. look here https://deploy-preview-28463--twbs-bootstrap.netlify.com/docs/4.3/components/buttons/#checkboxes-and-radio-buttons - these aren't "button groups" and reusing that class just to encapsulate the button-like "input followed by label" or "label wrapped around an input and some text acting as the visible label text" would be confusing and misleading, i'd say

ysds commented:
> fundamentally, it's confusing to just piggy-back onto btn-group for what is not an actual button group, but rather a button-like input. 

The current inline dropdown's markup is:

```html
<div class="btn-group">
  <button type="button" class="btn btn-primary dropdown-toggle" data-toggle="dropdown" ...>
    Action
  </button>
  <div class="dropdown-menu">
     ...
  </div>
</div>
```
(Color variant snippets in https://getbootstrap.com/docs/4.3/components/dropdowns/#single-button)

What do you think about this? That is not an actual button group.

(I'm still supporting using `.btn-group` for dropdowns and toggle buttons)

patrickhlauke commented:
not a fan of the use for dropdowns either, but can more easily rationalise that in that it groups a button and a dropdown together. less so when it's grouping only one thing (and a hidden radio/checkbox)

patrickhlauke commented:
incidentally, the `.btn-input-wrapper` class i proposed here could be renamed to something more generic like `.btn-wrapper`, and applying that to those dropdowns would have the same effect as well.

or perhaps taking a step back, renaming `.btn-group` to `.btn-wrapper`, and patching in any extra styles/selectors created here for `.btn-input-wrapper` and applying them to that, could also work. gives the class itself a more generic name, and it's then applied in different contexts where it "wraps" various things relating to a button-like look.

ysds commented:
I don't stick to the name of the `.btn-group`. I just don't want to increase the complexity of selectors and CSS. The changes to the `_button-group.scss` are very complex for now.

Looking for a wrapper class name (alt `.btn-group`) that everyone is convinced might be a great solution.

patrickhlauke commented:
btw note that even if we just reused `.btn-group` it wouldn't remove the need for *all* of the new selectors (like `.btn-input`), as they cover scenarios that weren't covered before just through `.btn-group`

patrickhlauke commented:
@mdo let me know if you want me to have a crack at renaming `btn-group` (and what your preference would be, if any) and re-simplifying some of the css selectors?

patrickhlauke committed:
Merge branch 'master' into v5-input-btn-group

ysds commented:
> btw note that even if we just reused .btn-group it wouldn't remove the need for all of the new selectors (like .btn-input)

Yep, I showed that (just adding one new class) in my original PR #25185.

patrickhlauke committed:
Merge branch 'master' into v5-input-btn-group

Johann-S commented:
Hey @twbs/css-review and @twbs/js-review  I removed QUnit and switch to Jasmine, so now your unit tests should go there: https://github.com/twbs/bootstrap/blob/master/js/src/button/button.spec.js

js/tests/unit/button.js have been removed, so if you need any help do not hesitate 😉

patrickhlauke commented:
@mdo just bumping this to see if it's likely to be used/picked up/worked on further...

patrickhlauke commented:
so...coming back to this, as i worked my socks off at the time to reorganise stuff, create some more explanations and examples, etc ... what's happening? should this all be closed? is it too far gone to salvage the bits that aren't controversial (since the creation/renaming of the class wasn't to everybody's liking)? /cc @mdo

patrickhlauke commented:
so... 5 months later, and still tumbleweed on this? makes me regret having put time and effort into trying to rationalise/improve stuff here to be honest...

XhmikosR commented:
Who's up to do a proper rebase for this branch? 😇

patrickhlauke commented:
cursed branch :)

ffoodd commented:
Just discovering this one! Looks interesting but pretty confused with all the things going on here, so if anyone's able to summarize the need, I'd be glad to help.

And if not, this topic might not be clear enough!

patrickhlauke commented:
in very broad strokes, from what i remember, this change does the following:

- kills off the clunky JS-based button groups for checkbox/radio button groups, instead relying purely on using the underlying `<input type="radio">` / `<input type="checkbox">` ... avoiding a LOT of the extra messing around we have had with these components in the past (all to make them behave like real radios/checkboxes)
- renames (currently, though that was then a big point of contention) the button group class
- reorders and expands examples to clarify and more clearly distinguish between the *styles* to make button groups and then the actual dynamic component to handle ARIA-based toggles

ffoodd commented:
If I understood correctly, we have three features that kind of interact with each others:
1. styling a button group — without any specific behaviour, just to style multiple appended buttons of any kind;
2. first behaviour: using `radio`s or `checkbox`es;
2. second behaviour: using JS and `aria-pressed`.

And after reading this dicussion carefully, I have a question: the first need for this PR was to remove JS. I can't see anyone in this discussion arguing to keep this JS part (except the fact that it already exists…). Am I wrong?

@patrickhlauke IMHO both behaviours are to be accomplished with `input`s.

Is there any reason to keep mentionning the ARIA pattern here? Am I missing something?

patrickhlauke commented:
ARIA is still needed for toggle buttons (`<button aria-pressed="true">`). faking them using, say, an `<input type="checkbox">` is not ideal/recommended.

patrickhlauke commented:
not to keep flogging a dead horse, but... @mdo et al, any idea on this whether we want to proceed or not?

MartijnCuppens commented:
@patrickhlauke, I'll have a look in another PR with another approach. So you're ok with dropping the js think with active classes and just rely on the CSS only label-buttons for v5?

patrickhlauke commented:
yes, the move to CSS only (except for controls/toggles that need `aria-pressed`) is non-controversial. but this PR also contains reordering/reorganisation of some docs aspects (moving the radio button group stuff, making up new classes for those situations, and lots more specific usage examples). it's that stuff as well that i'm quite keen to get a decision on

mdo commented:
Closing for #30650.
